File: isegoria/src/main/java/com/eulersbridge/isegoria/poll/PollsFragment.java
package com.eulersbridge.isegoria.poll;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.UiThread;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.eulersbridge.isegoria.MainActivity;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.network.api.models.Poll;
import com.eulersbridge.isegoria.util.Constants;
import com.eulersbridge.isegoria.util.ui.SimpleFragmentPagerAdapter;
import com.eulersbridge.isegoria.util.ui.TitledFragment;

import org.parceler.Parcels;

import java.util.List;
import java.util.Vector;

public class PollsFragment extends Fragment implements TitledFragment, MainActivity.TabbedFragment {

    private TabLayout tabLayout;
    private ViewPager viewPager;
    private SimpleFragmentPagerAdapter pagerAdapter;

	private List<Fragment> fragments;

    @Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View rootView = inflater.inflate(R.layout.poll_vote_fragment, container, false);

        fragments = new Vector<>();

        // Ensure options menu from another fragment is not carried over
		if (getActivity() != null)
            getActivity().invalidateOptionsMenu();

        PollsViewModel viewModel = ViewModelProviders.of(this).get(PollsViewModel.class);
		viewModel.getPolls().observe(this, polls -> {
		    if (polls != null)
		        addPolls(polls);
        });

        setupViewPager(rootView);
        setupTabLayout();
		
		return rootView;
	}

    @Override
    public String getTitle(Context context) {
        return context.getString(R.string.section_title_poll);
    }

    private void setupViewPager(View rootView) {
        if (rootView == null) rootView = getView();

        if (viewPager == null && rootView != null) {
            viewPager = rootView.findViewById(R.id.poll_vote_view_pager);

            pagerAdapter = new SimpleFragmentPagerAdapter(getChildFragmentManager(), fragments) {
                @Override
                public CharSequence getPageTitle(int position) {
                    return getString(R.string.poll_title, position + 1);
                }
            };
            viewPager.setAdapter(pagerAdapter);

            viewPager.setCurrentItem(0);
        }
    }

    @Override
    public void setupTabLayout(TabLayout tabLayout) {
        this.tabLayout = tabLayout;

        tabLayout.removeAllTabs();
        tabLayout.setVisibility(View.VISIBLE);
    }

    @UiThread
    private void updateTabs() {
        pagerAdapter.notifyDataSetChanged();

        tabLayout.setVisibility(fragments.size() < 2? View.GONE : View.VISIBLE);
    }

    @Override
    public void onPause() {
        super.onPause();

        if (tabLayout != null) tabLayout.removeOnTabSelectedListener(onTabSelectedListener);
    }

    private final TabLayout.OnTabSelectedListener onTabSelectedListener = new TabLayout.OnTabSelectedListener() {
        @Override
        public void onTabSelected(TabLayout.Tab tab) {
            viewPager.setCurrentItem(tab.getPosition());
        }

        @Override
        public void onTabUnselected(TabLayout.Tab tab) { }

        @Override
        public void onTabReselected(TabLayout.Tab tab) { }
    };

    private void setupTabLayout() {
        if (tabLayout == null) return;

        tabLayout.setupWithViewPager(viewPager);
        tabLayout.addOnTabSelectedListener(onTabSelectedListener);
    }

    private void addPolls(@NonNull List<Poll> polls) {
        if (getActivity() != null) {
            getActivity().runOnUiThread(() -> {
                for (Poll poll : polls) {
                    PollVoteFragment pollVoteFragment = new PollVoteFragment();

                    Bundle args = new Bundle();
                    args.putParcelable(Constants.ACTIVITY_EXTRA_POLL, Parcels.wrap(poll));

                    pollVoteFragment.setArguments(args);

                    fragments.add(pollVoteFragment);
                    updateTabs();
                }
            });
        }
	}
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/poll/PollsFragment.kt
package com.eulersbridge.isegoria.poll

import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.annotation.UiThread
import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.core.view.isVisible
import com.eulersbridge.isegoria.MainActivity
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.network.api.model.Poll
import com.eulersbridge.isegoria.util.extension.observe
import com.eulersbridge.isegoria.util.extension.runOnUiThread
import com.eulersbridge.isegoria.util.ui.SimpleFragmentPagerAdapter
import com.eulersbridge.isegoria.util.ui.TitledFragment
import dagger.android.support.AndroidSupportInjection
import kotlinx.android.synthetic.main.poll_vote_fragment.*
import java.util.*
import javax.inject.Inject

class PollsFragment : Fragment(), TitledFragment, MainActivity.TabbedFragment {

    @Inject
    internal lateinit var modelFactory: ViewModelProvider.Factory
    private lateinit var viewModel: PollsViewModel

    private var tabLayout: TabLayout? = null
    private lateinit var pagerAdapter: SimpleFragmentPagerAdapter
    private var fragments: MutableList<Fragment> = Vector()

    private val onTabSelectedListener = object : TabLayout.OnTabSelectedListener {
        override fun onTabSelected(tab: TabLayout.Tab) {
            viewPager.currentItem = tab.position
        }

        override fun onTabUnselected(tab: TabLayout.Tab) {}
        override fun onTabReselected(tab: TabLayout.Tab) {}
    }

    override fun onAttach(context: Context?) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
        viewModel = ViewModelProviders.of(this, modelFactory)[PollsViewModel::class.java]
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val rootView = inflater.inflate(R.layout.poll_vote_fragment, container, false)

        // Ensure options menu from another fragment is not carried over
        activity?.invalidateOptionsMenu()

        observe(viewModel.getPolls()) { polls ->
            if (polls != null)
                addPolls(polls)
        }

        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        setupViewPager()
    }

    override fun getTitle(context: Context?) = context?.getString(R.string.section_title_poll)

    private fun setupViewPager() {
        pagerAdapter = object : SimpleFragmentPagerAdapter(childFragmentManager, fragments) {
            override fun getPageTitle(position: Int)
                    = getString(R.string.poll_title, position + 1)
        }

        viewPager.apply {
            adapter = pagerAdapter
            currentItem = 0
        }
    }

    override fun setupTabLayout(tabLayout: TabLayout) {
        this.tabLayout = tabLayout

        tabLayout.apply {
            removeAllTabs()
            visibility = View.VISIBLE

            setupWithViewPager(viewPager)
            addOnTabSelectedListener(onTabSelectedListener)
        }
    }

    @UiThread
    private fun updateTabs() {
        pagerAdapter.notifyDataSetChanged()

        tabLayout?.isVisible = fragments.size >= 2
    }

    override fun onPause() {
        tabLayout?.removeOnTabSelectedListener(onTabSelectedListener)
        super.onPause()
    }

    private fun addPolls(polls: List<Poll>) {
        val newFragments = polls.map {
            PollVoteFragment().apply {
                arguments = bundleOf(FRAGMENT_EXTRA_POLL to it)
            }
        }

        fragments.addAll(newFragments)

        runOnUiThread {
            updateTabs()
        }
    }
}